src/renderer/components/+namespaces/namespace.store.ts

```diff
@@ -21,32 +21,103 @@ export const namespaceUrlParam = createPageParam<string[]>({
 export class NamespaceStore extends KubeObjectStore<Namespace> {
   api = namespacesApi;
-  contextNs = observable.array<string>();
+
+  @observable contextNs = observable.array<string>();
+  @observable isReady = false;
+
+  whenReady = when(() => this.isReady);
 
   constructor() {
     super();
     this.init();
   }
 
-  private init() {
-    this.setContext(this.initNamespaces);
-
-    return reaction(() => this.contextNs.toJS(), namespaces => {
-      storage.set(namespaces); // save to local-storage
-      namespaceUrlParam.set(namespaces, { replaceHistory: true }); // update url
-    }, {
-      fireImmediately: true,
+  onContextChange(callback: (contextNamespaces: string[]) => void, opts: IReactionOptions = {}): IReactionDisposer {
+    return reaction(() => this.contextNs.toJS(), callback, {
       equals: comparer.identity,
+      ...opts,
     });
   }
 
+  private async init() {
+    await clusterStore.whenLoaded;
+    if (!getHostedCluster()) return;
+    await getHostedCluster().whenReady; // wait for cluster-state from main
+    this.isReady = true;
+
+    this.setContext(this.initNamespaces);
+
+    const disposers: IReactionDisposer[] = [];
+
+    // save selected namespaces to local-storage and update URL
+    disposers.push(
+      this.onContextChange(namespaces => {
+        storage.set(namespaces);
+        namespaceUrlParam.set(namespaces, { replaceHistory: true });
+      }, {
+        fireImmediately: true,
+        equals: comparer.identity,
+      })
+    );
+    
+    // auto-load allowed namespaces
+    disposers.push(
+      reaction(() => this.allowedNamespaces, () => this.loadAll(), {
+        fireImmediately: true,
+        equals: comparer.identity,
+      })
+    );
+
+    return disposers;
+  }
+
+  get allowedNamespaces(): string[] {
+    return toJS(getHostedCluster().allowedNamespaces);
+  }
+
   get initNamespaces() {
-    return namespaceUrlParam.get();
+    const allowedNamespaces = new Set(this.allowedNamespaces);
+    const lastUsedNamespaces = new Set(storage.get());
+
+    // remove previously saved, but currently disallowed namespaces
+    lastUsedNamespaces.forEach(namespace => {
+      if (!allowedNamespaces.has(namespace)) {
+        lastUsedNamespaces.delete(namespace);
+      }
+    });
+
+    // return previously saved and currently allowed namespaces
+    if (lastUsedNamespaces.size) {
+      return Array.from(lastUsedNamespaces);
+    }
+    // otherwise select "default" or first allowed namespace
+    else {
+      if (allowedNamespaces.has("default")) {
+        return ["default"];
+      } else if (allowedNamespaces.size) {
+        return [Array.from(allowedNamespaces)[0]];
+      }
+    }
+
+    return [];
+  }
+
+  getContextNamespaces(): string[] {
+    const namespaces = this.contextNs.toJS();
+
+    if (!namespaces.length) {
+      return [...this.allowedNamespaces]; // show all namespaces when nothing selected
+    }
+
+    return namespaces;
   }
 
+  /**
+   * @deprecated
```

Comment:
It's not used anywhere in the codebase, but because `NamespaceStore` is exported to `extension-api` removing it  would be breaking change.. I would suggest to remove without deprecation..